Feat/unstructured mesh (#53)
exasim-project/NeoFOAM: doc/basics/executor.rst

> .. _api_executor:

Reviewer: We have already an executor documentation in `doc/api/executor.rst` we should merge both.

Executor
========

Overview
^^^^^^^^

> NeoFOAM uses the MPI+X approach for parallelism, where X is the execution space used for parallelism. The `Executor` class uses Kokkos and provides an interface for memory management and specifics were to execute the operations:

Reviewer: ```suggestion
NeoFOAM uses the MPI+X approach for parallelism, where X is the execution space used for device parallelism. The `Executor` class uses Kokkos and provides an interface for memory management and specifics were to execute the operations:
```

- `CPUExecutor`: run on the CPU with MPI
- `OMPExecutor`: run on the CPU with OpenMP and MPI
- `GPUExecutor`: run on the GPU with MPI

Design
^^^^^^

> One of the design goals is the ability to quickly switch between the executor models at runtime.

Reviewer: ```suggestion
One of the design goals is the ability to easily switch between different execution spaces at run time, i.e. enabling the ability to switch between CPU and GPU execution without having to recompile NeoFOAM. This is achieved by passing the executor as an argument to the container types like Field as shown below. 

```

.. code-block:: cpp

        NeoFOAM::GPUExecutor gpuExec {};
        NeoFOAM::CPUExecutor cpuExec {};

        NeoFOAM::Field<NeoFOAM::scalar> GPUField(gpuExec, 10);
        NeoFOAM::Field<NeoFOAM::scalar> CPUField(cpuExec, 10);

> The `Executor` is a std::variant

Reviewer: ```suggestion
The `Executor` is a `std::variant <https://en.cppreference.com/w/cpp/utility/variant>`_ 
```

.. code-block:: cpp

    using executor = std::variant<OMPExecutor, GPUExecutor, CPUExecutor>;

> and allows to switch between the different strategies for memory allocation and execution at runtime. We use `std::visit` to switch between the different strategies:

Reviewer: ```suggestion
and allows to switch between the different strategies for memory allocation and execution at runtime. We use `std::visit <https://en.cppreference.com/w/cpp/utility/variant/visit>`_ to switch between the different strategies:
```
